### ISSUE TITLE
base theme has too much stuff

### ISSUE THREAD

**Issue by gwax** [getnikola/nikola#2744]
### Environment (if reporting a bug)

**Nikola Version:** 7.8.4

### Description:

The `base` theme uses a number of bootstrap specific css classes and html tag conventions that are not necessarily suitable / desirable for use in building out new themes. In my usage, building out a [skeleton](http://getskeleton.com/) based theme, I found that I had to fork the `base` theme and strip a bunch of stuff out, before I could properly derive my own theme.

I'd be happy to unify my efforts with mainline `base` and push out a PR that further strips things down. Would other people be interested in seeing this?

**Comment by ralsina** [comment]
Since some people are actually using base in their sites (yes, really) we would need to keep base working and do a "lower level" theme below it. Would you agree to work on that?

**Comment by Kwpolska** [comment]
I’m going to say that this is best left as-is. The aim of the `base` theme is to provide many things:

* generic templates for common components that most people don’t need/want to override (eg. comments)
* templates for other important pages, that we *expect* the majority of themes to override (eg. `base*.tmpl`)
* a minimum, basic theme that doesn’t look bad, and that can be customized with little or no template code changes (eg. the `slidemenu` theme — only adds CSS, JS, and ~one line of HTML)

When building a theme based on something else (a framework, or an existing template), there are things the designer of said pre-existing asset expects from the HTML code. The `base` theme also makes a set of assumptions — some of which might have been stolen from Bootstrap, but others are there to facilitate creating themes. And if the framework wants a very specific layout, especially if the layout is non-standard, you must override the templates anyway.

The thing is, you can’t easily draw the line with things like those. What constitutes a “standard” layout you can use, which will work for all (or at least 95% of) frameworks? What if there shows up a framework that needs another change the “ultra-base” theme did not account for?

Taking that into account, I believe that we should stick to the status quo, and if someone thinks that `base` does too much, they can just override it from within their themes.

**Comment by gour** [comment]
Let me just add my .02$ to the discussion since in the recent time I've been playing with Hugo a bit which has plenty of available themes available...

Here is interesting quote from its [docs](http://gohugo.io/themes/overview/):

> Hugo provides a robust theming system which is simple, yet capable of producing even the most complicated websites.
>The Hugo community has created a wide variety of beautiful themes, as demoed at themes.gohugo.io, ready for using in your own site.
> Hugo themes have been designed to be the perfect balance between simplicity and functionality. Hugo themes are powered by the excellent Go template library. If you are new to Go templates, see our primer on Go templates.
>Hugo themes support all modern features you come to expect. They are structured in such a way to eliminate code duplication. Themes are also designed to be very easy to customize while retaining the ability to maintain upgradeability as the upstream theme changes.
> Hugo currently doesn’t ship with a “default” theme, allowing the user to pick whichever theme best suits their project.

So, there is no 'default' or _base_ theme in Hugo and installing a new one is just a question of running _git clone theme's url_ within _themes_ directory.

I'd also question "...a minimum, basic theme that doesn’t look bad, and that can be customized with little or no template code changes..." - just compare [this](https://themes.getnikola.com/) gallery with [this](http://themes.gohugo.io/) one?

I believe that we can agree that themes are one of the Nikola's major Achilles' heel making new themes more complicated than it could be, so, maybe, one approach to fix it would be to simply not having 'default' theme or, as originally suggested. strip it down to the bare minimum?

However, I'm not overly familiar with Hugo's theming and can't say whether it is possible to not have 'default' _base_ theme without too much surgery...but, in any case, this is probably 8.0 theme (pun intended) anyway. ;)
I agree, that in most cases the difference is not very deep, but the potential new users would probably choose the latter application although Nikola is (probably) capable to produce same result as Hugo.

**Comment by gwax** [comment]
> The aim of the base theme is to provide many things:

> * generic templates for common components that most people don’t need/want to override (eg. comments)

I agree.

> * templates for other important pages, that we expect the majority of themes to override (eg. `base*.tmpl`)

Also agreee,.

> * a minimum, basic theme that doesn’t look bad, and

If its purpose is to be a _base_ for other themes, why does it need to not look bad? Wouldn't not looking bad be the purpose of a _minimal_ or "basic" theme?

> that can be customized with little or no template code changes (eg. the slidemenu theme — only adds CSS, JS, and ~one line of HTML)

Completely agree and this is the crux of my criticism of the existing `base` theme. There are a number of things in `base` that exist specifically for bootstrap3 (ex. `sr-only1). In order to customize with "little or no template code changes", I forked base and removed extraneous things that got in my way or pulled in stuff that I don't want.

**Comment by gwax** [comment]
> we would need to keep base working and do a "lower level" theme below it. Would you agree to work on that?

I'd be happy to work on that; it's mostly there already. If not "base" what would that "lower level" theme be called?

**Comment by felixfontein** [comment]
> I believe that we can agree that themes are one of the Nikola's major Achilles' heel making new themes more complicated than it could be, so, maybe, one approach to fix it would be to simply not having 'default' theme or, as originally suggested. strip it down to the bare minimum?

Well, nobody forces you to derive your theme from `base` or `base-jinja`. You can also create a theme from scratch without a parent. (That's what I've been doing for my blogs' themes.)

**Comment by Kwpolska** [comment]
@gwax 

> * a minimum, basic theme that doesn’t look bad, and
If its purpose is to be a base for other themes, why does it need to not look bad? Wouldn't not looking bad be the purpose of a minimal or "basic" theme?

The `theme.css` file in base can be overridden pretty easily.

> Completely agree and this is the crux of my criticism of the existing base theme. There are a number of things in base that exist specifically for bootstrap3 (ex. sr-only). In order to customize with "little or no template code changes", I forked base and removed extraneous things that got in my way or pulled in stuff that I don't want.

Could you provide a list of those things? Regarding `.sr-only`, I disagree. It was introduced for accessibility reasons, and the class being `.sr-only` is a direct result of stealing CSS code from Bootstrap (three simple rules)

---

@felixfontein

> Well, nobody forces you to derive your theme from base or base-jinja. You can also create a theme from scratch without a parent. (That's what I've been doing for my blogs' themes.)

I wouldn’t recommend it, although you can do something like “copy files from base until the site builds”, it’s easier to just override stuff you dislike in a child theme. (Themes without a parent are not allowed on themes.getnikola.com for compatibility and maintainability reasons).

---

@gour 

> So, there is no 'default' or base theme in Hugo and installing a new one is just a question of running git clone theme's url within themes directory.

`nikola theme -i nameOfTheme` is equally simple. And if a theme is not in the index, it’s also a matter of putting a folder in `themes/`.

> I'd also question "...a minimum, basic theme that doesn’t look bad, and that can be customized with little or no template code changes..." - just compare this gallery with this one? [Hugo vs Nikola]

False. How many designers do you know that tried to make a Nikola theme but resigned because it’s hard? Try comparing the a scale: Nikola has 1198 stars on GitHub, while Hugo has 16907. They have a larger community. Hugo has 16× more stars than Nikola and 4× more themes (154 vs 36).

I’m not an artist, and @ralsina isn’t either. To make a good theme, you mainly need *design* skills.

> I believe that we can agree that themes are one of the Nikola's major Achilles' heel making new themes more complicated than it could be, so, maybe, one approach to fix it would be to simply not having 'default' theme or, as originally suggested. strip it down to the bare minimum?

How many themes did you make? What problems did you encounter? If a designer would want to make a theme, they could always ask for help in the IRC, mailing list, or wherever.

---

Let’s talk about WordPress for a moment. I built a WordPress theme, almost from scratch, a year ago. I took html5blank and Bootstrap. It was a royal pain:

* Some CSS rules are **mandatory** (.wp-caption, align stuff, .screen-reader-text — duplicate of `.sr-only` from Bootstrap), and yet there’s an ugly <style> block in HTML code
* Menus are created using a weird framework, which has really ugly classes, and **requires** <li> tags to be used
* I had to hack together a custom plugin to make sidebars look good
* I had to copy-paste many CSS rules from Bootstrap to adhere to WordPress’s structures

And yet **everyone loves WordPress** and there are millions of themes for it.

**Comment by gour** [comment]
@Kwpolska 

> nikola theme -i nameOfTheme is equally simple.

I fully agree...my point was that installing Hugo theme is also simple procedure **despite** of not having mechanism for inheriting from some _base_ theme.

> False. How many designers do you know that tried to make a Nikola theme but resigned because it’s hard? 

No idea, but I speak based on numbers...

> Try comparing the a scale: Nikola has 1198 stars on GitHub, while Hugo has 16907. They have a larger community. 

Well, Nikola is (much) longer around than Hugo...one can explain that Go language is, in general, hyped at the moment and therefore has attracted  3x more (478 vs 159) contributors.

Otoh, I strongly believe that 14x more (16924 vs 1198) stars has much to do with the themes...

> Hugo has 16× more stars than Nikola and 4× more themes (154 vs 36).

Hmmm. 36 Nikola themes is the total number, 25 are listed as Mako-based and only 11 ones are based on Jinja2.  Moreover, most of those themes are inheriting from Bootstrap - that's the original point in this thread, iow. to make _base_ stripped from Bootstrap-stuff - which makes the look of the whole gallery  'less colourful'.

>I’m not an artist, and @ralsina isn’t either. To make a good theme, you mainly need design skills.

You and @ralsina are great developers and have made Nikola one of the best static-site-generators in existence - Hugo, with much bigger community, only recently got support for multi-lingual sites - a feature which is present in Nikola since time immemorial. Search is also available in Nikola in many flavours, while Hugo forum is still filled with so many posts in regard and only in last release there is mechanism to get custom/local search more easily. Same thing with e.g. support for commenting systems - Hugo is, out of the box, still Disqus-only.

We can continue with deploying support, plugins, ability to use so many markup formats...so, Nikola is, without a question, much richer feature-wise and therefore I doubt that Hugo's speed is the **only** reason why it has become much more popular...
 
Finally, based on the above, I strongly believe that number of themes is something which greatly contributes (@ralsina confirmed several times that the themes are the top-priority complain from people) that Nikola is not as popular as it deserves to. 

So, please take my input only as an attempt to help make Niikola's community bigger and/or stronger...for now I'll try to contribute by providing some themes.

> How many themes did you make?  What problems did you encounter? 

None, so far.

> If a designer would want to make a theme, they could always ask for help in the IRC, mailing list, or wherever.

I'm not designer as well and will ask for all the help I can get...

**Comment by Kwpolska** [comment]
> I fully agree...my point was that installing Hugo theme is also simple procedure despite of not having mechanism for inheriting from some base theme.

That’s not what the base theme is for! Its purpose is to make *creating* themes easier, not installing them.

---

I wanted to see the scale of “bootstrapisms”. I used the following regexes (yes, I know, parsing HTML with regex is evil, but lxml would choke on Mako)

```py
re.findall('class="(.*?)"', all_base_templates_text)
re.findall('\.([a-zA-Z-_][a-zA-Z0-9-_]*)', bootstrap_css)
```

A bit of cleanup and set magic, and we get this:

```py
{'active',  # Used for menus (<li class="active">). The most obvious name.
 'breadcrumb',  # Used for breadcrumbs, a pretty basic component.
 'carousel',  # Bootstrapism. Used for slides, which are a rather obscure feature.
 'carousel-control',
 'carousel-indicators',
 'carousel-inner',
 'disabled',  # For archive navigation, obvious name/coincidence.
 'hidden-print',   # For hiding comments and page navigation when printing. Stylistic choice.
 'item',  # Slides.
 'left',  # Pagers. Obvious.
 'next',
 'pager',
 'previous',
 'right',
 'sr-only',  # For accessibility (“(active)” markers for navbar; “Skip to main content”)
 'sr-only-focusable',
 'thumbnail'}  # Used in galleries.
```

Are there really that many? `hidden-print` missing won’t affect much; `sr-only` is a good thing to have, and the others are generally coincidences caused by reusing obvious names, or slides being obscure and heavily CSS+JS-dependent (and you really don’t want any carousel CSS in `base`). There are also CSS bits we expect every theme to have (maybe we should list them on our theming page), for example `.metadata`. However, it’s simple to change a theme to use something else, just by overriding templates with whatever you need.

**Comment by gwax** [comment]
It's worth distinguishing between accessibility features, which are definitely a good thing, and bootstrap's implementation of accessibility features (e.g. `.sr-only`), which don't _do_ anything unless you use or implement bootstrap's accessibility css/js.

If accessibility is the goal, we might as well support [WAI-ARIA](https://www.w3.org/WAI/intro/aria) ([2](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA)) ([3](https://www.w3.org/TR/wai-aria/)), rather than leave bootstrap accessibility classes in `base`. This probably falls out of scope of this issue and was already discussed in #377 (but might be worth revisiting).

**Comment by ralsina** [comment]
@gwax one way to make your point would be to just make such a theme. I will get it in themes.getnikola.com even without a parent, and all.

I have done a bunch of theme ports in the past few weeks, and I can say: doing a theme for Nikola is very easy.

* Create an empty theme (we should have a helper command for this, I'll make an issue)
* nikola theme -c base.tmpl
* nikola theme -c index.tmpl
* nikola theme -c post.tmpl
* Tweak each of those 3. If you seen base going into helpers and such, just ignore it and do your thing in base.tmpl instead.
* Use nikola auto a lot

Usually that is all it takes, depending on your design skills. Mine suck, so I just copy stuff, but I know that is all that's needed to achieve themes as dissimilar as [canterville](https://themes.getnikola.com/v7/canterville/demo/) and [maupassant](https://themes.getnikola.com/v7/maupassant/demo/)

**Comment by gwax** [comment]
@ralsina I've already built a stripped down _base_ theme and derived my own personal theme from it. I was thinking of contributing back some of my clean-ups and fixes to nikola **base** but I want to make sure it will be received well before I put the effort into making the PR.

I opened this issue as a reference point and for discussion in anticipation of creating the aforementioned PR, not as a request for someone else to fix something.

To that end, my questions remain:

- Would people appreciate a more spartan **base**?
- Should I contribute some sort of _core_ theme and re-derive **base** from it?
- Should I close this issue and just keep going my own?
- Would it be better to defer and get involved in refactoring the core themes whenever things move toward v8?

**Comment by Kwpolska** [comment]
> Create an empty theme (we should have a helper command for this, I'll make an issue)

`nikola theme -n foobar [--engine=mako] [--parent=base]`

@gwax: It would help if you made a list of changes you would make. How spartan is *spartan*?

**Comment by ralsina** [comment]
@gwax the guideline for base was "a very basic theme that looks decent" ... I suspect it went a little overboard on looking decent and is not basic enough.

Most contributions are well received, I would love to have your "core" theme in themes.getnikola.com and we can start working from there. I am not going to say "we'll base everything off it" because I am trying not to promise stuff anymore, and I just don't know what it looks like :-)

**Comment by ralsina** [comment]
@Kwpolska hey, I did not know that existed!

**Comment by gwax** [comment]
I have introduced getnikola/nikola-themes#120 with the changed portions of my more _austere_ base theme.

One of the things that irked me about the built-in base, which I just re-noticed is that all assets will be built and pulled-in, even if they aren't used by a child theme.

**Comment by gwax** [comment]
Ok, instead, I have started working on putting a **core** theme in nikola proper.

Here's my current progress: https://github.com/getnikola/nikola/compare/master...gwax:core_theme

Remaining steps on my end:

- Implement **core-jinja**
- Update documentation

From there, I would be happy to either:

1) open a PR against nikola
2) add themes to nikola-themes
3) use this as a replacement for **base**

Please let me know your initial reactions.

**Comment by Kwpolska** [comment]
> Implement core-jinja

You can automate that by using `scripts/jinjify.py`, but this should be done only when we accept the current theme.

Looking at this, I disagree with some changes. `rst.css` has to stay named as-is. Note that we need to keep as much backwards compatibility as we can. And as it is, **every** theme expects `rst.css` to be there. Same with dropping `messages_cz.py` and breaking some symlinks, as well as using non-minified assets. **You may not delete any macro, any file from the base theme.** This also restricts `core` in some places, because `base` should override as few files as possible.

**Comment by gwax** [comment]
I have done/will do:

- Change `rst.css` back to be named as such.
- Restore `messages_cz.py` (it was dropped by accident)
- Restore macros and templates that did not appear to be in use, to maintain backwards compatibility.
- Symlinks in base will stay in base.

Nikola provides filters that minify javascript and css, so I'm a little confused about the reason for putting minimized assets in the themes. It seems like one would want un-minimized assets in the theme and minimized assets in the generated output.

**Comment by Kwpolska** [comment]
No, nobody really wants un-minified assets. Those filters are not a default feature, and are hard to discover. Un-minified versions are useful only for people modifying those assets, of which I doubt there are many — and those people can already figure minification out for themselves.

**Comment by ralsina** [comment]
@gwax I usually when am making a theme work with unminified assets and then minimize them when finished.

**Comment by gwax** [comment]
As we seem to be arguing over details more than principles, I have opened #2810 with the work in progress.

**Comment by Kwpolska** [comment]
I think it’s solved in #2815.

**Comment by gwax** [comment]
I think the combination of #2812 and #2815 is probably enough to call this "good enough"
